terraform/eval_state.go

```diff
@@ -473,3 +484,44 @@ func (n *EvalForgetResourceState) Eval(ctx EvalContext) (interface{}, error) {
 
 	return nil, nil
 }
+
+// EvalRefreshDependencies is an EvalNode implementation that appends any newly
+// found dependencies to those saved in the state. The existing dependencies
+// are retained, as they may be missing from the config, and will be required
+// for the updates and destroys during the next apply.
+type EvalRefreshDependencies struct {
```

Comment:
That's a good point, and should have added a little more context as I would like to retain the state-only-refresh idea.  The only reason that refresh handling the "fix" was left in this PR was for TFC. TFC does not complete an apply if the plan shows no changes, hence has no way to fix the dependencies in the state without first applying a change.  Granted the use case of a "TFC config that will encounter destroy order issues without first applying a change" is rather small, so I can refactor the dependency handling out of refresh if desired.

Reply:
After trying out this idea without refresh writing out dependencies, I think it's probably worth the slightly odd idea of the refreshed state being altered by the config, even if it's not really the instance state itself. Since refresh is the only moment we have to re-write resource state (apply does not write state if there's no changes), if we don't do this existing resources would never have the depends_on field fixed up.